Финальное задание спринта №6

KanID92 opened the pull request:
HistoryManager:
 - добавить void remove(int id);

   InMemoryHistoryManager:
 - добавить реализацию void remove(int id);
 - добавить связный список на основе HashMap для истории;
 - убрать ограничение по количеству сохраненных задач в истории.

   InMemoryHistoryManagerTest:
 - добавить тесты.

   Main:
 - добавить пользовательский сценарий.

   Общее:
 - добавить файл api-tests.yml;

KanID92 committed:
Финальное задание спринта №6

   HistoryManager:
 - добавить void remove(int id);

   InMemoryHistoryManager:
 - добавить реализацию void remove(int id);
 - добавить связный список на основе HashMap для истории;
 - убрать ограничение по количеству сохраненных задач в истории.

   InMemoryHistoryManagerTest:
 - добавить тесты.

   Main:
 - добавить пользовательский сценарий.

   Общее:
 - добавить файл api-tests.yml;

kesch9 commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,22 +1,24 @@
-            }
-            lastTasksList.addLast(task);
-
+        if (task == null) {
```
Нужно переделать реализацию historyManager и написать свою реализацию связанного списка, также добавить удаление задач из истории

KanID92 replied:
Прошу прощения: из-за необходимости перезалить коммит, из-за автоматической проверки которая длилась больше часа - откатил слишком далеко (была запушена версия без изменений).  Восстановил заново верную версию и запушил: коммит bbfc2c3

kesch9 reviewed (commented):
Задание не выполнено

KanID92 committed:
Финальное задание спринта №6 v.2

ПОДГРУЖЕНА ВЕРНАЯ ВЕРСИЯ

   HistoryManager:
 - добавить void remove(int id);

   InMemoryHistoryManager:
 - добавить реализацию void remove(int id);
 - добавить связный список на основе HashMap для истории;
 - убрать ограничение по количеству сохраненных задач в истории.

   InMemoryHistoryManagerTest:
 - добавить тесты.

   Main:
 - добавить пользовательский сценарий.

   Общее:
 - добавить файл api-tests.yml;

kesch9 commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,93 @@
+    }
 
+    @Override
+    public void add(Task task) {
```
Обязательно
Не нужно было удалять проверку на null входного параметра, нужно вернуть

kesch9 commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,93 @@
     @Override
+    public void remove(int id) {
+        Node node = history.get(id);
+        removeNode(node);
```
Обязательно
Нужно перед удалением проверить что node не null и такая запись есть

kesch9 commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,93 @@
+    Node first;
+    Node last;
+
+    void linkLast(Task task) {
```
Обязательно
Метод делаем private

kesch9 commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,93 @@
+        removeNode(node);
+    }
+
+    public List<Task> getAll() {
```
Обязательно
Внутренние методы, которые не в интерфейсе, делаем private и можно было не выносить в отдельный метод getAll, чтобы вызвать его в getHistory, а написать именно там реализацию

kesch9 commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,93 @@
+        history.remove(node.data.getId());
+    }
+
+    public Task getLast() {
```
Обязательно
Этого метода не должно быть по заданию, в тестах вы же можете получить историю и посмотреть последний элемент, у вас же отдается в нужно порядке и последний элемент будет равен last

kesch9 commented on src/service/InMemoryTaskManager.java:
```diff
@@ -148,17 +145,20 @@ public void updateSubtask(SubTask subTask) { // Обновление подза
     @Override
     public void deleteAllTasks() {  //удаление всех обычных задач
-        tasks.clear();
+        tasks.clear(); // удаление самих задач;
```
Обязательно
При удалении всех задач удаляются задачи из истории

kesch9 commented on src/service/InMemoryTaskManager.java:
```diff
@@ -225,8 +229,21 @@ private void changeEpicProgress(int epicId) {
+        }
+    }
+
+    public HistoryManager getHistoryManager() { // для тестов
```
Обязательно
Этого метода не должно быть согласно заданию и ради тестов писать его не надо, в тестах работаем через taskManager, например у вас вот так написано
taskManager.getHistoryManager().remove(subTaskTest1Id);
переписать можно вот так
taskManager.deleteSubtaskByID(subTaskTest1Id);

kesch9 commented on src/service/TaskManager.java:
```diff
@@ -49,4 +48,7 @@ public interface TaskManager {
 
     List<Task> getHistory();
+
+    HistoryManager getHistoryManager(); // для тестов
```
Обязательно
Удаляем этот метод из интерфейса

kesch9 commented on test/service/InMemoryHistoryManagerTest.java:
```diff
@@ -1,50 +1,120 @@
+        subTaskTest1Id = taskManager.createSubtask(new SubTask("Задача №5", "Для тестов", epicTest1Id));
+        subTaskTest1 = taskManager.getSubtaskByID((subTaskTest1Id)); // вызов 5 задачи
+
+        taskManager.getTaskByID(1); // вызов 1 задачи
```
Можно лучше
Зачем хардкодить id если у вас есть поля с id, например вот здесь taskTest1Id

kesch9 commented on test/service/InMemoryHistoryManagerTest.java:
```diff
@@ -1,50 +1,120 @@
     TaskManager taskManager = Managers.getDefault();
     int taskTest1Id;
     int taskTest2Id;
+    int taskTest3Id;
```
Можно лучше
Выносить id в отдельные переменные не нужно, вам нужны только сами задачи, вы у них можете получить id

kesch9 commented on test/service/InMemoryTaskManagerTest.java:
```diff
@@ -17,7 +14,7 @@ class InMemoryTaskManagerTest {
         taskManager.createTask(new Task("Задача №1", "Задача №1 для проверки"));
         taskManager.createEpic(new Epic("Эпик №2", "Эпик №2 для проверки"));
-        taskManager.createSubtask(new SubTask("Подзадача №3", "Подзадача №3 для проверки",2));
+        taskManager.createSubtask(new SubTask("Подзадача №3", "Подзадача №3 для проверки", 2));
```
Обязательно
нет тестов согласно заданию
_Удаляемые подзадачи не должны хранить внутри себя старые id.
Внутри эпиков не должно оставаться неактуальных id подзадач._
и еще в задании предлагается подумать над этим вопросом
_С помощью сеттеров экземпляры задач позволяют изменить любое своё поле, но это может повлиять на данные внутри менеджера. Протестируйте эти кейсы и подумайте над возможными вариантами решения проблемы._

kesch9 reviewed (commented):
Есть проблемы при реализации

KanID92 committed:
Финальное задание спринта №6 (Корректировка по ревью):

class InMemoryHistoryManager:
- добавить проверку на null в методе add(Task task);
- добавить проверку node на null и на наличие в списке;
- изменить модификатор доступа к Node first и Node last и linkLast(Task task) на private;
- удалить метод getAll(), реализация метода перенесена в getHistory();
- удалить метод getLast().

class InMemoryHistoryManagerTest
- откорректировать тест shouldLinkLast() из-за удаления метода getLast();
- откорректировать тест shouldLinkLast() из-за удаления getHistoryManager();
- добавлен тест shouldNotContainInEpicDeletedSubtask();
- добавлен тест shouldNotContainInDeletedSubtaskEpicID();

interface HistoryManager:
- удалить метод Task getLast().

interface TaskManager:
- удалить метод getHistoryManager();

class InMemoryTaskManager:
- удалить метод getHistoryManager().

class Subtask:
- добавить метод deleteEpicId().

KanID92 commented:
Корректировка по ревью. Коммит f7792d0.

kesch9 commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,89 @@
     @Override
+    public void remove(int id) {
+        Node node = history.get(id);
+        if (node != null && history.containsValue(node)) {
```
Можно лучше
Вторая проверка не нужна
history.containsValue(node)
если нет, то вы получите null и вы это проверяете первым условием.
Достаточно
if (node != null)

kesch9 commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,89 @@
+
     public List<Task> getHistory() {
-        return new LinkedList<>(lastTasksList);
+        ArrayList<Task> historyList = new ArrayList<>();
```
Забыли проставить над методом
@Override

kesch9 reviewed (approved):
Approve

kesch9 reviewed (approved):
Approve
